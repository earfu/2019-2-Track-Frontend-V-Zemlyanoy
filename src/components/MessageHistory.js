import React, { useEffect } from 'react';
// import { FlatList } from 'react-native';
import PropTypes from 'prop-types';
import MessageItem from './MessageItem';

function MessageHistory({ messageArray }) {
  // forced scroll to the end with every new message
  useEffect(() => document.querySelector('#last').scrollIntoView(false), [
    messageArray.length,
  ]);
  return (
    <div className="message-history-list">
      {messageArray.map((item, index) => (
        <li key={item.number}>
          <MessageItem
            author={item.author}
            date={item.date}
            number={item.number}
            text={item.text}
          />
        </li>
      ))}
      <div id="last" />
    </div>
  );
}

MessageHistory.propTypes = {
  messageArray: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default MessageHistory;
