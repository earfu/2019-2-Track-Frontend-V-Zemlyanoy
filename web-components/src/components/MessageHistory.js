import chatDefaults from '../chatDefaults';

const template = document.createElement('template');
template.innerHTML = `
    <style>
        .message-area {
            font: 12pt bold;
            display: flex;
            flex-direction: column;
            justify-content: flex-end;
        }
    </style>
    <div class="message-area"></div>
`;

export default class MessageHistory extends HTMLElement {
  constructor() {
    super();
    /* this.shadowRoot = */this.attachShadow({ mode: 'open' });
    this.shadowRoot.appendChild(template.content.cloneNode(true));
    this.$messageArea = this.shadowRoot.querySelector('.message-area');
    this.head = null;
    this.recreate();
  }

  recreate(chatName) { // clear history, read the localStorage, form the message list, append list
    this.clear(); // full reset; clearing is also done in writeList(), though
    const name = `${chatName || chatDefaults.firstChatName}`;
    const lsKey = `messages_${name}`;
    const lsString = localStorage.getItem(lsKey); // the messages stored
    if (lsString === null) {
      return;
    }

    let messageList = null;
    let pos = 0;
    const len = lsString.length;
    let sep1 = 0;
    let msgLength = 0;
    while (pos < len) {
      sep1 = lsString.indexOf('|', pos);
      msgLength = Number.parseInt(lsString.slice(pos, sep1), 10);
      const msgString = lsString.slice(pos, sep1 + msgLength + 1);
      const msg = document.createElement('message-item');
      msg.fromString(msgString);
      messageList = (messageList === null) ? msg : messageList.add(msg);
      // can change list head; also must watch out for list being a null
      pos = sep1 + msgLength + 1;
      // this list is assembled in linear time, assuming older messages at start
    }

    this.writeList(messageList);
  }

  append(msg) { // append a message to the history
    // note: no using msg.store() as long as this function is used in recreate()
    msg.setPrevious(this.head);
    this.head = msg;
    this.$messageArea.append(msg);
    msg.scrollIntoView(false);
  }

  writeList(listHead) { // write a full list from the head (i.e., end), overwriting past history
    this.clear(); // remove everything
    let current = listHead;
    const msgArea = this.$messageArea;
    while (current.previous !== null) {
      msgArea.prepend(current);
      current = current.previous;
    }
    msgArea.prepend(current);
    this.head = listHead;
    listHead.scrollIntoView(false); // set starting position at the newest
  }

  addList(listHead) { // for merging histories in date order
    let firstHead = this.head; // head of the list added to (i.e., the one already in the history)
    if (firstHead === null) { // as long as added list is in order; sort it here?
    // the MessageItem class currently has no methods for assembling lists without date-ordering;
    // MessageHistory does, however
      this.writeList(listHead);
      return;
    }
    let current = listHead;
    while (current !== null) {
      const temp = current.previous;
      firstHead = firstHead.add(current);
      current = temp; // current.previous may be changed by adding
    }

    this.writeList(firstHead); // to display messages in order
  }

  clear() {
    const msgArea = this.$messageArea;
    const messages = msgArea.children;
    const msgArray = [];
    for (let i = 0; i < messages.length; i += 1) {
      msgArray[i] = messages[i];
    }
    for (let i = 0; i < msgArray.length; i += 1) {
      msgArea.removeChild(msgArray[i]);
    }
    // no direct looping through messages,
    // as the collection itself is affected by removals
    this.head = null;
  }

  scrollEnd() {
    if (this.head === null) {
      return;
    }
    this.head.scrollIntoView(false);
  }
}

customElements.define('message-history', MessageHistory);
