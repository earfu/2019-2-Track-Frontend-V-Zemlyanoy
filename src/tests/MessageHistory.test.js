import React from 'react';
import renderer from 'react-test-renderer';
import MessageHistory from '../components/MessageHistory';

function messageFactory(messageArray) {
  const len = messageArray.length;
  const item = {
    number: len,
    author: `author_${len}`,
    text: `Message ${len} text content`,
    date: 1000000 + 10000 * len,
    image: null,
    audio: null,
  };
  return item;
}

it('renders correctly', () => {
  const messageArray = [];
  for (let i = 0; i < 10; i += 1) {
    messageArray.push(messageFactory(messageArray));
  }

  const component = renderer
    .create(<MessageHistory messageArray={messageArray} />)
    .toJSON();
  // empty function used; a warning here is fine, function calls are tested in Storybook
  expect(component).toMatchSnapshot();
});
