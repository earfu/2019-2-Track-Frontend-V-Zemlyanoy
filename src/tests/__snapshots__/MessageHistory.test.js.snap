// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="message-history-list"
>
  <li>
    <div
      className="message-item-div"
      id={0}
    >
      <p
        className="message-item-text"
      >
        Message 0 text content
      </p>
      <p
        className="message-item-author"
      >
        author_0
      </p>
      <p
        className="message-item-date"
      >
        1/1/1970, 12:16:40 AM
      </p>
    </div>
  </li>
  <li>
    <div
      className="message-item-div"
      id={1}
    >
      <p
        className="message-item-text"
      >
        Message 1 text content
      </p>
      <p
        className="message-item-author"
      >
        author_1
      </p>
      <p
        className="message-item-date"
      >
        1/1/1970, 12:16:50 AM
      </p>
    </div>
  </li>
  <li>
    <div
      className="message-item-div"
      id={2}
    >
      <p
        className="message-item-text"
      >
        Message 2 text content
      </p>
      <p
        className="message-item-author"
      >
        author_2
      </p>
      <p
        className="message-item-date"
      >
        1/1/1970, 12:17:00 AM
      </p>
    </div>
  </li>
  <li>
    <div
      className="message-item-div"
      id={3}
    >
      <p
        className="message-item-text"
      >
        Message 3 text content
      </p>
      <p
        className="message-item-author"
      >
        author_3
      </p>
      <p
        className="message-item-date"
      >
        1/1/1970, 12:17:10 AM
      </p>
    </div>
  </li>
  <li>
    <div
      className="message-item-div"
      id={4}
    >
      <p
        className="message-item-text"
      >
        Message 4 text content
      </p>
      <p
        className="message-item-author"
      >
        author_4
      </p>
      <p
        className="message-item-date"
      >
        1/1/1970, 12:17:20 AM
      </p>
    </div>
  </li>
  <li>
    <div
      className="message-item-div"
      id={5}
    >
      <p
        className="message-item-text"
      >
        Message 5 text content
      </p>
      <p
        className="message-item-author"
      >
        author_5
      </p>
      <p
        className="message-item-date"
      >
        1/1/1970, 12:17:30 AM
      </p>
    </div>
  </li>
  <li>
    <div
      className="message-item-div"
      id={6}
    >
      <p
        className="message-item-text"
      >
        Message 6 text content
      </p>
      <p
        className="message-item-author"
      >
        author_6
      </p>
      <p
        className="message-item-date"
      >
        1/1/1970, 12:17:40 AM
      </p>
    </div>
  </li>
  <li>
    <div
      className="message-item-div"
      id={7}
    >
      <p
        className="message-item-text"
      >
        Message 7 text content
      </p>
      <p
        className="message-item-author"
      >
        author_7
      </p>
      <p
        className="message-item-date"
      >
        1/1/1970, 12:17:50 AM
      </p>
    </div>
  </li>
  <li>
    <div
      className="message-item-div"
      id={8}
    >
      <p
        className="message-item-text"
      >
        Message 8 text content
      </p>
      <p
        className="message-item-author"
      >
        author_8
      </p>
      <p
        className="message-item-date"
      >
        1/1/1970, 12:18:00 AM
      </p>
    </div>
  </li>
  <li>
    <div
      className="message-item-div"
      id={9}
    >
      <p
        className="message-item-text"
      >
        Message 9 text content
      </p>
      <p
        className="message-item-author"
      >
        author_9
      </p>
      <p
        className="message-item-date"
      >
        1/1/1970, 12:18:10 AM
      </p>
    </div>
  </li>
  <div
    id="last"
  />
</div>
`;
